File: Data Structures & Algorithms/max-area-of-island/submission-2.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        ROWS, COLS = len(grid), len(grid[0])
        
        visited = set()
        def dfs(r, c):
            if min(r, c) < 0 or r >= ROWS or c >= COLS or (r, c) in visited or grid[r][c] == 0:
                return 0
            
            visited.add((r, c))
            up = dfs(r-1, c)
            down = dfs(r+1, c)
            left = dfs(r, c-1)
            right = dfs(r, c+1)

            return 1 + up + down + left + right

        maxArea = 0
        for i in range(ROWS):
            for j in range(COLS):
                if grid[i][j] == 1:
                    maxArea = max(dfs(i, j), maxArea)

        return maxArea

File: Data Structures & Algorithms/minimum-path-sum/submission-2.cpp
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int ROWS = grid.size(), COLS = grid[0].size();
        vector<int> dp(COLS);

        int total = 0;
        for (int i{COLS - 1}; i >= 0; --i) {
            total += grid[ROWS - 1][i];
            dp[i] = total;
        }

        for (int r = ROWS - 2; r >= 0; --r) {
            vector<int> nextDp(COLS);
            nextDp[COLS - 1] = dp[COLS - 1] + grid[r][COLS - 1];
            for (int c = COLS - 2; c >= 0; --c) {
                nextDp[c] = min(nextDp[c + 1], dp[c]) + grid[r][c];
            }
            swap(nextDp, dp);
        }

        return dp[0];
    }
};